src/python/pants/backend/tools/yamllint/rules.py

```diff
@@ -0,0 +1,185 @@
+# Copyright 2022 Pants project contributors (see CONTRIBUTORS.md).
+# Licensed under the Apache License, Version 2.0 (see LICENSE).
+from __future__ import annotations
+
+import os
+from collections import defaultdict
+from dataclasses import dataclass
+from typing import Any, cast
+
+from pants.backend.python.util_rules.pex import Pex, PexProcess, PexRequest
+from pants.backend.tools.yamllint.subsystem import Yamllint
+from pants.core.goals.lint import LintFilesRequest, LintResult
+from pants.core.goals.tailor import group_by_dir
+from pants.core.util_rules.partitions import Partition, Partitions
+from pants.engine.fs import Digest, DigestSubset, MergeDigests, PathGlobs
+from pants.engine.internals.native_engine import FilespecMatcher, Snapshot
+from pants.engine.process import FallibleProcessResult
+from pants.engine.rules import Get, MultiGet, collect_rules, rule
+from pants.util.frozendict import FrozenDict
+from pants.util.logging import LogLevel
+from pants.util.strutil import pluralize
+
+
+class YamllintRequest(LintFilesRequest):
+    tool_subsystem = Yamllint
+
+
+@dataclass(frozen=True)
+class PartitionInfo:
+    config_snapshot: Snapshot | None
+
+    @property
+    def description(self) -> str:
+        if self.config_snapshot:
+            return self.config_snapshot.files[0]
+        else:
+            return "<default>"
+
+
+@dataclass(frozen=True)
+class YamllintConfigFilesRequest:
+    filepaths: tuple[str, ...]
+
+
+@dataclass(frozen=True)
+class YamllintConfigFiles:
+    snapshot: Snapshot
+    source_dir_to_config_files: FrozenDict[str, str]
+
+
+def find_nearest_ancestor_file(files: set[str], dir: str, config_file: str) -> str | None:
+    while True:
+        candidate_config_file_path = os.path.join(dir, config_file)
+        if candidate_config_file_path in files:
+            return candidate_config_file_path
+
+        if dir == "":
+            return None
+        dir = os.path.dirname(dir)
+
+
+@rule
+async def gather_config_files(
```

Comment:
Can you add a `@TODO` here (and to `scalafmt`) that these implementations are similar and should likely be genericized and shared? You don't have to address it, just jot it down.